cmd/av/stack_sync.go

```diff
@@ -0,0 +1,280 @@
+package main
+
+import (
+	"emperror.dev/errors"
+	"encoding/json"
+	"fmt"
+	"github.com/aviator-co/av/internal/git"
+	"github.com/aviator-co/av/internal/stacks"
+	"github.com/kr/text"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+)
+
+// stackSyncConfig contains the configuration for a sync operation.
+// It is serializable to JSON to handle the case where the sync is interrupted
+// by a merge conflict (so it can be resumed with the --continue flag).
+type stackSyncConfig struct {
+	// Set the parent of the current branch to this branch.
+	// This effectively re-roots the stack on a new parent (e.g., adds a branch
+	// to the stack).
+	Parent string `json:"parent"`
+	// If set, only sync up to the current branch (do not sync descendants).
+	// This is useful for syncing changes from a parent branch in case the
+	// current branch needs to be updated before continuing the sync.
+	Current bool `json:"current"`
+	// If set, incorporate changes from the trunk (repo base branch) into the stack.
+	// Only valid if synchronizing the root of a stack.
+	// This effectively re-roots the stack on the latest commit from the trunk.
+	Trunk bool `json:"trunk"`
+	// If set, do not push to GitHub.
+	NoPush bool `json:"noPush"`
+	// If set, we need to continue the current sync step before syncing the
+	// remaining branches.
+	// Not serialized to JSON because it's only set by the --continue flag.
+	Continue bool `json:"-"`
+}
+
+// stackSyncState is the state of an in-progress sync operation.
+// It is written to a file if the sync is interrupted (so it can be resumed with
+// the --continue flag).
+type stackSyncState struct {
+	CurrentBranch string          `json:"currentBranch"`
+	Config        stackSyncConfig `json:"config"`
+}
+
+var stackSyncFlags struct {
+	// Include all the options from stackSyncConfig
+	stackSyncConfig
+	// If set, we're continuing a previous sync.
+	Continue bool
+	// If set, abort an in-progress sync operation.
+	Abort bool
+}
+
+var stackSyncCmd = &cobra.Command{
+	Use:   "sync",
+	Short: "synchronize stacked branches",
+	Long: strings.TrimSpace(`
+Synchronize stacked branches to be up-to-date with their parent branches.
+
+By default, this command will sync all branches starting at the root of the
+stack and recursively rebasing each branch based on the latest commit from the
+parent branch.
+
+If the --current flag is given, this command will not recursively sync dependent
+branches of the current branch within the stack. This allows you to make changes
+to the current branch before syncing the rest of the stack.
+
+If the --trunk flag is given, this command will synchronize changes from the
```

Comment:
Yeah, I agree (though we can workshop the flag name a bit). Let's do that in a later PR. 😛 (I'm thinking maybe `--latest-trunk` or `--to-trunk` rather than `--with-trunk`)

Reply:
Sure we can workshop the name, but I think is syncing from trunk and not to trunk.